pageserver/src/tenant/timeline.rs

```diff
@@ -3946,13 +3956,26 @@ impl Timeline {
                 LayerResidenceStatus::Resident,
                 LayerResidenceEventReason::LayerCreate,
             );
-            self.insert_historic_layer(x, &mut updates, mapping);
+            if mapping.contains(&x) {
+                duplicated_layers.insert(x.layer_desc().key());
+                warn!(
+                    "duplicated layers generated in compaction: {:?}",
+                    x.layer_desc().key() // TODO: change to use Display
+                );
+            } else {
+                self.insert_historic_layer(x, &mut updates, mapping);
+            }
         }
 
         // Now that we have reshuffled the data to set of new delta layers, we can
         // delete the old ones
         let mut layer_names_to_delete = Vec::with_capacity(deltas_to_compact.len());
         for l in deltas_to_compact {
+            if duplicated_layers.contains(&l.key()) {
+                // skip duplicated layers, they will not be removed; we have already overwritten them
+                // with new layers in the compaction phase 1.
+                continue;
```

Comment:
Is this preparation for the L0 => L0 compaction or why does this happen? Surely the L1 we overwrote is not part of the `deltas_to_compact`?

I find the `l.key()` check here vs. earlier `duplicated_layers.insert(x.layer_desc().key());` (R3960) confusing. If the key is available both ways, why have the other?

Reply:
`l` here is `LayerDesc`, renamed to `ldesc` to make it more clear in https://github.com/neondatabase/neon/pull/4696/commits/610b38e3a28035626b417b957338959b5afbde88

Reply:
Aah oki ... 610b38e is fine I guess.

How about the my original question? Had I misunderstood something there as well?
> Is this preparation for the L0 => L0 compaction or why could this check be needed? Surely the L1 we overwrote is not part of the deltas_to_compact (L0s)?

Reply:
Even for L0 -> L1 compaction now I can think of a case of producing exactly the same file name. If L0 happens to be 256MB perfectly sized files and they are not overlapping, compaction will produce exactly the same set of file. In this case, we should do this check and not removing the original file.

Reply:
Aah so in the case of somehow producing L0s which would be the same as the L1s which they'd compact to, as in, produce L0s with the same partitioning.. Yeah I guess this is possible in the general case. Transaction related data probably makes this practically impossible.

Perhaps this protection could be mentioned in a comment, as far as I remember there was none.

Reply:
I've updated the comment in the code for this case and the previous companion failure case around the fail point we are adding for integration test in https://github.com/neondatabase/neon/pull/4696/commits/53d0df80f42367fde0352f2caf78acbef41df434

Reply:
I think the added failpoint now is a good ugly indicator of this!

One thing I realized over the weekend is that:

- L0s will be 0x00...00-0xff..ff regardless of even having keys that go up to u128::MAX
- so if we would be able to manufacture the L0s in question, they would still be named:
  - 00..00_ff..ff-(ealier_lsn)_(later_lsn)

As such, the L0s given as input and the L1s produced as output would always be distinct as the definition is that L1 does not cover the full key range. The only way could happen is that compaction somehow produced L0s (outside of L0 => L0 compaction).

While noting the above, I still think your finding is valid *until* we have an assertion that none of the compact results are recognizable as "needing more L0 compaction". As in, we should have an `assert!(!LayerMap::is_l0(...))` for each produced delta layer. Otherwise we'd create an infinite "needs compaction" state.

If you agree, this conv is good to resolve. If you think I misunderstood something, please point it out to me in slack.

Reply:
In practice, I don't see Postgres generating keys starting with 0x000 / 0xfff, so I think we will never have this infinite compaction problem, and it would be good to add this assertion.

Reply:
fixed in https://github.com/neondatabase/neon/pull/4696/commits/1a077f682cb004d1a78f374405ee83ee739752bd, it's an error and compactor will no longer schedule jobs if errors are detected, so we don't need to panic/assert here.